src/librustc_mir/borrow_check/type_check/mod.rs

```diff
@@ -1275,7 +1287,13 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
                             opaque_decl.concrete_ty, resolved_ty, renumbered_opaque_defn_ty,
                         );
 
-                        if !concrete_is_opaque {
+                        if !concrete_is_opaque
+                            || (is_function_return
+                                && matches!(opaque_decl.origin, hir::OpaqueTyOrigin::FnReturn))
```

Comment:
Hmm... It feels ungreat that desugaring information is used for semantic purposes?

Reply:
It's already load bearing in `member_constraint_feature_gate` and `visit_opaque_types`.